keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,410 @@
+   cache re-initialization.
+
+[Clone()]: https://pkg.go.dev/github.com/google/btree#BTree.Clone
+
+### Cache Inconsistency detection
+
+For the first iteration of the mechanism we will calculate the hash based on LIST response
+at some specific RV of each resource and expose it as a metric. Hash will be calculated for
+both cache and etcd, allowing users to setup an alert to detect if they don't match.
+
+Once every 5 minutes (default compaction interval) we will send a LIST request to both etcd and
+watch cache and calculate hash of the response.
+
+TODO: Decide whether to calculate hash based on object fields or values stored in etcd.
+
+### Risks and Mitigations
+
+#### Memory overhead
+
+B-tree snapshots are designed to minimize memory overhead by storing pointers to
+the actual objects, rather than the objects themselves. Since the objects are
+already cached to serve watch events, the primary memory impact comes from the
+B-tree structure itself. To quantify the memory overhead, we run 5k scalability tests.
+They should represent the worst case scenario, as they utilize large number of small objects.
+The results are promising:
+
+* **Object Allocations:** Allocation profile collected during the test test has
+  shown an increase of 7GB in object allocations, which translates to a
+  negligible 0.2% of total allocations.
+* **Memory Usage:** Memory in use profile collected during the test has shown
+  Btree memory usage of 300MB, representing a 1.3% of total memory used.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+- Ensure the pagination is well tested
+
+##### Unit tests
+
+- `k8s/apiserver/pkg/storage/cache`: `2024-12-12` - `<test coverage>`
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+Given we're only modifying kube-apiserver, integration tests are sufficient.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature gate
+- Feature is covered with unit and integration tests
+
+
+#### Beta
```

Comment:
please include that the mismatch metric has not failed in alpha testing.

Reply:
Done